### ISSUE TITLE
Threads not cleaned up after an authentication failure

### ISSUE THREAD

**Issue by darrenfoong** [paramiko/paramiko#2214]
**Environment**

- Debian 12 (bookworm)
- Python 3.11.2
- Paramiko 2.12.0

**Problem**

- `client.connnect()` does not clean up threads after an authentication failure, e.g. wrong password, despite running `client.close()`.
- This can be replicated with [this file](https://github.com/darrenfoong/paramiko-expts/blob/main/main.py); comment out lines 82-84 accordingly; instructions in lines 7-8.
- With a correct password, the number of threads remains at 1.
- With a wrong password, the number of threads is 6 (1 + 5 attempts?)
- I tried to run an internal function of `Transport` to clean up the threads, but it seems sensitive to timing, so probably a race condition exists.

I've searched for similar issues; although there are some about memory leaks (e.g. #949), I couldn't find any about thread leaks, nor any about leaks after an authentication failure.

**Comment by bskinn** [comment]
Thanks for the report, @darrenfoong!

Based on my understanding of what changed in moving to the 3.0 series, I doubt a fix for this entered the codebase with the 3.0 release... but, to probe the shape of the problem a bit, could you test with the following combinations?

- Python 3.9.x, Paramiko 2.12.0
- Python 3.11.2, Paramiko 3.1.x
- Python 3.9.x, Paramiko 3.1.x

I'm especially curious if this thread misbehavior could be something that emerged due to the performance work that's gone into 3.10 and 3.11.

Thanks!

**Comment by darrenfoong** [comment]
I've tested the following combinations:

 - Python 3.9.16, Paramiko 2.12.0
 - Python 3.11.2, Paramiko 3.1.0
 - Python 3.9.16, Paramiko 3.1.0

and they all have the same results:

```
correct password - before run - num of threads: 1
correct password - after run - num of threads: 1

wrong password - before run - num of threads: 1
wrong password - after run - num of threads: 6

wrong password with stop_thread - before run - num of threads: 1
wrong password with stop_thread - after run - num of threads: 2
```

**Comment by jun66j5** [comment]
That is caused by not calling `client.close()` when authentication failing.

At least, you should call `client.close()` (or use `with client`) regardless of the authentication result.

```diff
     for _ in range(0, 5):
-        try:
-            run(client, "wrong_password")
-            client.close()
-        except:
-            pass
+        with client:
+            try:
+                run(client, "wrong_password")
+                client.close()
+            except:
+                pass

```

**Comment by darrenfoong** [comment]
@jun66j5 you are right! I updated the code and the threads are cleaned up, albeit after some time. It was my mistake for not calling `client.close()` in a `finally` block or, as you suggested, using a `with` statement.

I'm actually investigating a resource leak in a bigger code base that uses Fabric, and I thought I found an issue in Paramiko. Turns out that it's in Fabric instead (https://github.com/fabric/fabric/blob/2.7.1/fabric/connection.py#L688)

```
    def close(self):
        if self.is_connected:
            self.client.close()
            if self.forward_agent and self._agent_handler is not None:
                self._agent_handler.close()
```

When authentication fails, `self.is_connected` is `False` and the internal Paramiko client will not be closed.

I've updated my code to test Fabric instead, and will report an issue to the Fabric team.
